tests/test_render.py

```diff
@@ -2275,3 +2281,265 @@ def set_bc(a):
+                ),
+                message='Failed assertion: #{}.value != {}'.format(
+                    id, assertions[id]
+                )
+            )
+
+
+    def test_simple_clientside_serverside_callback(self):
+        app = dash.Dash(__name__, assets_folder='test_clientside')
+
+        app.layout = html.Div([
+            dcc.Input(id='input'),
+            html.Div(id='output-clientside'),
+            html.Div(id='output-serverside')
+        ])
+
+
+        @app.callback(
+            Output('output-serverside', 'children'),
+            [Input('input', 'value')])
+        def update_output(value):
+            return 'Server says "{}"'.format(value)
+
+
+        app.clientside_callback(
+            ClientsideFunction(
+                namespace='clientside',
+                function_name='display'
+            ),
+            Output('output-clientside', 'children'),
+            [Input('input', 'value')]
+        )
+
+        self.startServer(app)
+
+        input = WebDriverWait(self.driver, 10).until(
+            EC.presence_of_element_located((By.ID, "input"))
+        )
+        WebDriverWait(self.driver, 10).until(
+            EC.text_to_be_present_in_element(
+                (By.ID, 'output-serverside'),
+                'Server says "None"'
+            )
+        )
+
+        WebDriverWait(self.driver, 10).until(
+            EC.text_to_be_present_in_element(
+                (By.ID, 'output-clientside'),
+                'Client says "undefined"'
+            )
+        )
+
+        input.send_keys('hello world')
+        WebDriverWait(self.driver, 10).until(
+            EC.text_to_be_present_in_element(
+                (By.ID, 'output-serverside'),
+                'Server says "hello world"'
+            )
+        )
+        WebDriverWait(self.driver, 10).until(
+            EC.text_to_be_present_in_element(
+                (By.ID, 'output-clientside'),
+                'Client says "hello world"'
+            )
+        )
+
+
+    def test_chained_serverside_clientside_callbacks(self):
+        app = dash.Dash(__name__, assets_folder='test_clientside')
+
+        app.layout = html.Div([
+
+            html.Label('x'),
+            dcc.Input(id='x', value=3),
+
+            html.Label('y'),
+            dcc.Input(id='y', value=6),
+
+            # clientside
+            html.Label('x + y (clientside)'),
+            dcc.Input(id='x+y'),
+
+            # server-side
+            html.Label('x+y / 2 (serverside)'),
+            dcc.Input(id='x+y / 2'),
+
+            # server-side
+            html.Div([
+                html.Label('Display x, y, x+y/2 (serverside)'),
+                dcc.Textarea(id='display-all-of-the-values'),
+            ]),
+
+            # clientside
+            html.Label('Mean(x, y, x+y, x+y/2) (clientside)'),
+            dcc.Input(id='mean-of-all-values'),
+
+        ])
+
+        app.clientside_callback(
+            ClientsideFunction('R', 'add'),
```

Comment:
As discussed, I think we should isolate clienside callbacks in a predictable dash-property/container in the windows object to prevent arbitrary calls on arbitrary global `window`` object / functions.

Reply:
Ah, OK I actually don't understand this. So the user wouldn't write `window.their_namespace`, they would write `window.dash_clientside.their_namespace`?

Reply:
Yes. I would suggest that the users JS entrypoint be something like `window.dash_clientside` and ClientsideFunction(namespace, function_name) result in `window.dash_clientside[namespace][function_name]` being called.

Reply:
Or we could expose a function like
```js
registerClientside('myNamespace', {
    myFunc: (a, b) => a + b
})
```
That might also be nice later on for validation of function references - though I guess we'd get that from `window.dash_clientside` as well - but with a registration function it would be easier to ensure that multiple calls wouldn't clobber each other (ie the user writes something like `window.dash_clientside = {namespace: {...} }` in two different files)

Reply:
It is nice allowing access to ramda out of the box (both as a clientside namespace on its own and off `window` so you can use it from your own clientside functions)... how did it get to be `window.R` though? I feel like it'd be better as `ramda`, at least for its clientside namespace, especially with dashR coming up.

Reply:
> how did it get to be window.R though? I feel like it'd be better as ramda, at least for its clientside namespace, especially with dashR coming up.

it's actually not here out of the box! I included the ramda bundle from their CDN and put it in the test `assets/`

Reply:
I think I'll start with just regular namespaces, then do register in a follow up PR

Reply:
> it's actually not here out of the box! I included the ramda bundle from their CDN and put it in the test assets/

Ah I see, `window.R` is how the CDN version exports itself... OK, as long as that's not added by dash itself that's fine.

> I think I'll start with just regular namespaces, then do register in a follow up PR

Sure, the boilerplate you need right now is a bit steep but it should be easy to add a register fn later in a backward-compatible way.